cmd/soroban-rpc/internal/methods/get_transactions.go

```diff
@@ -134,92 +116,141 @@ func (h transactionsRPCHandler) getTransactionsByLedgerSequence(ctx context.Cont
-		for i := startTxIdx; i <= txCount; i++ {
-			cursor = toid.New(int32(ledger.LedgerSequence()), int32(i), 1)
+		*txns = append(*txns, txInfo)
+		if len(*txns) >= int(limit) {
+			break
+		}
+	}
 
-			ingestTx, err := reader.Read()
-			if err != nil {
-				if errors.Is(err, io.EOF) {
-					// No more transactions to read. Start from next ledger
-					break
-				}
-				return GetTransactionsResponse{}, &jrpc2.Error{
-					Code:    jrpc2.InvalidParams,
-					Message: err.Error(),
-				}
-			}
+	return cursor, nil
+}
 
-			tx, err := db.ParseTransaction(ledger, ingestTx)
-			if err != nil {
-				return GetTransactionsResponse{}, &jrpc2.Error{
-					Code:    jrpc2.InternalError,
-					Message: err.Error(),
-				}
-			}
+// getTransactionsByLedgerSequence fetches transactions between the start and end ledgers, inclusive of both.
+// The number of ledgers returned can be tuned using the pagination options - cursor and limit.
+func (h transactionsRPCHandler) getTransactionsByLedgerSequence(ctx context.Context,
+	request GetTransactionsRequest,
+) (GetTransactionsResponse, error) {
+	ledgerRange, err := h.ledgerReader.GetLedgerRange(ctx)
+	if err != nil {
+		return GetTransactionsResponse{}, &jrpc2.Error{
+			Code:    jrpc2.InternalError,
+			Message: err.Error(),
+		}
+	}
 
-			txInfo := TransactionInfo{
-				ApplicationOrder:    tx.ApplicationOrder,
-				FeeBump:             tx.FeeBump,
-				ResultXdr:           base64.StdEncoding.EncodeToString(tx.Result),
-				ResultMetaXdr:       base64.StdEncoding.EncodeToString(tx.Meta),
-				EnvelopeXdr:         base64.StdEncoding.EncodeToString(tx.Envelope),
-				DiagnosticEventsXDR: base64EncodeSlice(tx.Events),
-				Ledger:              tx.Ledger.Sequence,
-				LedgerCloseTime:     tx.Ledger.CloseTime,
-			}
-			txInfo.Status = TransactionStatusFailed
-			if tx.Successful {
-				txInfo.Status = TransactionStatusSuccess
-			}
+	err = request.isValid(h.maxLimit, ledgerRange)
+	if err != nil {
+		return GetTransactionsResponse{}, &jrpc2.Error{
+			Code:    jrpc2.InvalidRequest,
+			Message: err.Error(),
+		}
+	}
 
-			txns = append(txns, txInfo)
-			if len(txns) >= int(limit) {
-				break LedgerLoop
-			}
+	start, limit, err := h.initializePagination(request)
+	if err != nil {
+		return GetTransactionsResponse{}, err
+	}
+
+	// Iterate through each ledger and its transactions until limit or end range is reached.
+	// The latest ledger acts as the end ledger range for the request.
+	var txns []TransactionInfo
+	cursor := toid.New(0, 0, 0)
+LedgerLoop:
+	for ledgerSeq := start.LedgerSequence; ledgerSeq <= int32(ledgerRange.LastLedger.Sequence); ledgerSeq++ {
+		ledger, err := h.fetchLedgerData(ctx, uint32(ledgerSeq))
+		if err != nil {
+			return GetTransactionsResponse{}, err
+		}
+
+		cursor, err = h.processTransactionsInLedger(ledger, start, &txns, limit)
+		if err != nil {
+			return GetTransactionsResponse{}, err
+		}
+
+		if len(txns) >= int(limit) {
+			break LedgerLoop
```

Comment:
You can remove the tag, doing break by itself will lead to the same result.

Reply:
Ah nice catch. Forgot to remove this.